MSC2230: Store Identity Server in Account Data (#2230)
matrix-org/matrix-doc: proposals/2230-identity-server-account-data.md

# MSC2230: Store Identity Server in Account Data

The URL of the Identity Sever to use is currently specified at registration and
login time and then used for the lifetime of a login session. If users wish to
specify a custom one, they must do so each time they log in on every client.
Once they have chosen an Identity Server to advertise their 3PIDs on, it would
be normal that they would wish to continue using this Identity Server for all
Identity requests in their account accross all clients. This proposal aims to
make this easier.

## Proposal

The base URL of the Identity Server is to be stored in user account data. It
shall be stored in the same format as in a .well-known file under the key,
`m.identity_server` and shall comprise a single key, `base_url` which is the

> base URL of the ID Server to use (that is, the part before `/_matrix`).

Reviewer: Including https?

> Upon registration or login, a client MUST refrain from performing any requests

Reviewer 2: We have no framework for enforcing this:
```suggestion
Upon registration or login, a client SHOULD refrain from performing any requests
```

to the Identity Server until the account data has been fetched from the server.

> Once it has the account data, it MUST check for the presence of the

Reviewer 2: ```suggestion
Once it has the account data, it SHOULD check for the presence of the
```

> `m.identity_server` key. If present, the `base_url` in this key MUST be used

Reviewer 2: ```suggestion
`m.identity_server` key. If present, the `base_url` in this key SHOULD be used
```

as the Identity Server base URL for the duration of the login session. If this
key is not present, the client SHOULD populate it with the ID Server URL

> that was or would have been used in the login/registration process. This may

Reviewer 2: "may" can imply that there's a limited set of options. To reduce ambiguity, we can use a different word:
```suggestion
that was or would have been used in the login/registration process. This could
```

be either from user input, a .well-known lookup, or a default in the client.

> Client MUST listen for changes in the `m.identity_server` account data value

Reviewer 2: ```suggestion
Client SHOULD listen for changes in the `m.identity_server` account data value
```

> and update the URL that they use for ID Server requests accordingly UNLESS

Reviewer 2: UNLESS is not a keyword we use. I would recommend just lowercasing it.
```suggestion
and update the URL that they use for ID Server requests accordingly unless
```

the login session and choice of ID Server base URL predates this change, in
which case they SHOULD continue to use the value they are currently using.

> Clients MAY offer a way for users to change the ID server being used. If they

Reviewer 2: MAY is not a keyword we use:
```suggestion
Clients can offer a way for users to change the ID server being used. If they
```

# MSC2230: Store Identity Server in Account Data

> The URL of the Identity Sever to use is currently specified at registration and

Reviewer 3: ```suggestion
The URL of the Identity Server to use is currently specified at registration and
```

login time and then used for the lifetime of a login session. If users wish to
specify a custom one, they must do so each time they log in on every client.
Once they have chosen an Identity Server to advertise their 3PIDs on, it would
be normal that they would wish to continue using this Identity Server for all
Identity requests in their account accross all clients. This proposal aims to
make this easier.

## Proposal

The base URL of the Identity Server is to be stored in user account data. It

> shall be stored in the same format as in a .well-known file under the key,

Reviewer 3: can you give an example of what this will look like?

Reviewer 2: ```
{
  "type": "m.identity_server",
  "content": {
    "base_url": "https://vector.im"
  }
}
```

if I understand the docs correctly, I think.

> shall be stored in the same format as in a .well-known file under the key,

Reviewer 3: ```suggestion
shall be stored in the same format as in a .well-known file under the event type
```

`m.identity_server` and shall comprise a single key, `base_url` which is the
base URL of the ID Server to use (that is, the part before `/_matrix`, including
`https://`).

Upon registration or login, a client SHOULD refrain from performing any requests
to the Identity Server until the account data has been fetched from the server.
Once it has the account data, it SHOULD check for the presence of the
`m.identity_server` key. If present, the `base_url` in this key SHOULD be used
as the Identity Server base URL for the duration of the login session. If this

> key is not present, the client SHOULD populate it with the ID Server URL

Reviewer 4: After some discussion in the privacy sync (recorded in https://github.com/vector-im/riot-web/issues/10597), it seems easier and reasonable for clients to not write any IS in account data in login / startup.

that was or would have been used in the login/registration process. This could
be either from user input, a .well-known lookup, or a default in the client.

Client SHOULD listen for changes in the `m.identity_server` account data value
and update the URL that they use for ID Server requests accordingly unless
the login session and choice of ID Server base URL predates this change, in

> which case they SHOULD continue to use the value they are currently using.

Reviewer 4: After some discussion in the privacy sync, it seems easier and reasonable for clients to apply any changes via account data as they come down. (From https://github.com/vector-im/riot-web/issues/10597)

Clients can offer a way for users to change the ID server being used. If they
do, the client MUST update the value of `m.identity_server` accordingly.

> The `m.identity_server` may be present with a value of `null`. In this case,

Reviewer 3: account data is a list of events (which  inherently have a content dict) rather a set of completely arbitrary values, so do you mean:

```suggestion
The `m.identity_server` may be present with a `base_url` of `null`. In this case,
```

clients MUST treat this as no ID Server URL being set and not perform ID
Server requests, disabling any functionality that requires such requests.

Conversely, if a user wishes to disable ID Server functionality, the client
shall action this by setting the `base_url` of the `m.identity_server`
account data entry to `null`.

### Transition Period

Clients currently logged in with a value configured for the ID Server base
URL SHOULD use the value from `m.identity_server` if it exists or is created,
but otherwise continue to use the URL they had previously. They MUST NOT
populate the `m.identity_server` with their current ID Server base URL.

## Tradeoffs

There are a number of ways to transition to this new scheme. Clients could
populate the account data with their current ID Server URL as soon as
possible, and immediately use any new value seen on account data. This
would a much faster migration but any users with clients using different
ID Servers would suddenly find all their clients using the ID Server of
whichever client they updated first.

## Potential issues

Users will no longer be able to have different clients configured with
different ID Servers.

> ## Security considerations

Reviewer 3: do clients need to be a bit wary of things they read from the URL to check that they are valid URLs and aren't localhost and that sort of thing?

> An attacker would be able to force all a user clients to use a given ID Server

Reviewer 3: it also puts control of the IS to be used in the hands of the HS admin (or a MITM etc).

Reviewer 5: This could be solved by signing `m.identity_server` account data using the user's master cross-signing key.  (Once we get that through MSC...)
